In [14]:
acc_covmis_before = 0.9885245901639345
f1_covmis_before = 0.9843518094365552

acc_covmis_after = 0.9881862404447533
f1_covmis_after = 0.9827856884000605

(acc_covmis_before - acc_covmis_after), (f1_covmis_before - f1_covmis_after)


(0.0003383497191811191, 0.001566121036494672)

In [16]:
acc_liar2_before = 0.9378068739770867
f1_liar2_before = 0.9074319933338915

acc_liar2_after = 0.9324912891986062
f1_liar2_after = 0.9217986228840329

(acc_liar2_before - acc_liar2_after), (f1_liar2_before - f1_liar2_after)


(0.005315584778480487, -0.014366629550141452)

In [2]:
acc = 0.9885245901639345
human_baseline = 0.9254098360655738

(acc - human_baseline) * 100


6.311475409836065

In [6]:
x = {"query": "Below is a CLAIM and the PRIOR KNOWLEDGE associated with it. Please classify the CLAIM as TRUE or FALSE based on the PRIOR KNOWLEDGE. If the content described by the CLAIM is correct, then classify it as TRUE; if the content described by the CLAIM is incorrect, then classify it as FALSE.\n\nCLAIM: According to \"Article 116,\" a state of emergency will begin across Colombia at 6 a.m. on the 17th.\nPublication date: 2020-03-17\n\nPRIOR KNOWLEDGE:\nInformation 1:\nPublication date: None\nTitle: Emergency Provisions\nContent:\nIf approved by both the houses, the Emergency continues for 6 months and can be extended to an indefinite period with an approval of the Parliament for every six months.\nSuspension of Fundamental rights under Article 19: According to Article 358, when a proclamation of National Emergency is made, the six fundamental rights under article 19 are automatically suspended. Article 19 is automatically revived after the expiry of the emergency.\nArticle 356 empowers the President to issue a proclamation if he is satisfied that a situation has arisen in which the government of a state cannot be carried on in accordance with the provisions of the constitution.\nArticle 365 says that whenever a state fails to comply with or to give effect to any direction from the centre, it will be lawful for the President to hold that a situation has arisen in which the government of the state cannot be carried on in accordance with the provisions of the constitution.\nInformation 2:\nPublication date: None\nTitle: National Emergency\nContent:\nIf approved by both the houses, the Emergency continues for 6 months and can be extended to an indefinite period with an approval of the Parliament for every six months. Every resolution approving the proclamation of emergency or its continuance must be passed by either House of Parliament by a special majority.\nSuspension of Fundamental rights under Article 19: According to Article 358, when a proclamation of National Emergency is made, the six fundamental rights under article 19 are automatically suspended. Article 19 is automatically revived after the expiry of the emergency.\nEffect on fundamental rights: Articles 358 and 359 describes the effect of a National Emergency on the Fundamental Rights. These two provisions are explained below: Suspension of Fundamental rights under Article 19: According to Article 358, when a proclamation of National Emergency is made, the six fundamental rights under article 19 are automatically suspended.\nEven when the emergency was in operation, the third proclamation of National Emergency was made in June 1975. Both the second and the third proclamations were revoked in March 1977 · Article 355 imposes a duty on the centre to ensure that the government of every state is carried on in accordance with the provisions of the constitution.\nInformation 3:\nPublication date: 2024-06-07\nTitle: The Emergency (India) - Wikipedia\nContent:\nWith a wafer-thin majority of 7 to 6, the bench of the Supreme Court restricted Parliament's amendment power by stating it could not be used to alter the \"basic structure\" of the Constitution. Subsequently, Prime Minister Gandhi made A. N. Ray—the senior-most judge amongst those in the minority in Kesavananda Bharati—Chief Justice of India.\nThis was due to many reasons such as RBI's policy of putting in place a 6 per cent ceiling on annual money supply growth months before the emergency, record monsoon in the year of 1975 leading to record harvest of foodgrains which led to food prices declining, increased import of grains, and reduced demand due to cutting of worker's wages and bonuses.\nHowever, these reduced prices only lasted till March 1976 when the prices of commodities started to go up again, on account of foodgrain production declining by 7.9%. Between 1 April and 6 October 1976 the wholesale price index rose by 10%, in which the price of rice rose by 8.3%, groundnut oil rose by 48%, while the prices of industrial raw materials as a group rose by 29.3%.\nThe emergency regime exempted those earning between Rs 6,000-8,000 from taxation, provided tax breaks for those earning between Rs 8,000-15,000 in the range of Rs 45-264. There were only 3.8 million (38 lakh) taxpayers in the country at the time. Wealth taxes were also cut from 8% to 2.5% while the income taxes on those earning more than Rs 100,000 were reduced from 77% to 66%. This was expected to lower the government's revenue by Rs 3.08-3.25 billion.\nDuring Emergency SGPC offices used to be flooded with Anti-Emergency activists from RSS and BJP who had taken shelter to escape the government. According to Amnesty International, 140,000 people had been arrested without trial during the twenty months of Gandhi's Emergency. Jasjit Singh Grewal estimates that 43,000 of them came from India's two per cent Sikh minority.\nInformation 4:\nPublication date: 2020-02-01\nTitle: Emergency in India: Explanation of Article 352 - 360 under the Constitution\nContent:\nAccording to Article 354, the distribution of revenue between the Union and the State can be altered by the Centre. As per Article 83(2), the normal life of the Lok Sabha may be extended by the President by a year each time up to a period not exceeding 6 months after the proclamation ceases to operate.\nThe Proclamation will remain in operation for 6 months after it has been approved by the Parliament. The duration of an emergency can be extended for 6 months at a time but it cannot remain in operation for more than 3 years. By a subsequent Proclamation, a proclamation of State Emergency can be revoked.\nIt is the duty of the Union to ensure that the State remains protected from disturbance and external aggression, while the Proclamation of Emergency is in operation. The Union shall ensure that the State Government works according to the provisions of the Constitution.\nThe Union shall ensure that the State Government works according to the provisions of the Constitution. As per Article 356, if the President after receiving a report from the Governor of a State or otherwise is satisfied that such a situation exists where the Government of a State cannot be carried in accordance with the provisions of the Constitution, he may issue a Proclamation.\nInformation 5:\nPublication date: 2020-05-07\nTitle: Colombian President Duque declares new state of emergency to ease pain of virus lockdown | Reuters\nContent:\nBOGOTA (Reuters) - Colombia has declared a second state of emergency to support sectors of the economy that will remain shut down for an extended period to reduce the spread of the novel coronavirus, President Ivan Duque said on Wednesday.\nColombian fiscal authorities on Monday cleared the way for the government to widen its deficit to 6.1% of gross domestic product this year, owing to an expected 5.5% economic contraction.\nFILE PHOTO: Demonstrators take part in a protest demanding government food aid for the poor, amid the coronavirus disease (COVID-19) outbreak in Bogota, Colombia April 16, 2020. REUTERS/Luisa Gonzalez/File Photo Purchase Licensing Rights, opens new tab\nUnder the measures, the government will subsidize workers' pay to the tune of 40% of the minimum wage in companies whose turnover has fallen by at least 20% due to the impact of the coronavirus outbreak. The minimum wage in Colombia is just shy of 1 million pesos a month, or around $250.\n\nHere is a clear summary of the given INFORMATION:\n\n**Information 1 and 2**: These two pieces of information discuss the Emergency Provisions in the Indian Constitution. They explain that if a national emergency is declared, it can continue for 6 months and be extended indefinitely with parliamentary approval. During this time, certain fundamental rights under Article 19 are suspended, but are revived after the emergency ends. The President has the power to issue a proclamation of emergency if a situation arises where a state government cannot function according to the Constitution.\n\n**Information 3**: This article discusses the Emergency period in India in 1975-1977, during which time the government implemented various economic measures, including tax breaks and exemptions. The article also mentions that many people were arrested without trial during this period, including a significant number of Sikhs.\n\n**Information 4**: This article explains the provisions of Articles 352-360 of the Indian Constitution, which relate to emergency powers. It discusses how the President can declare a state of emergency, which can be extended for up to 3 years, and how the Union government is responsible for ensuring that state governments function according to the Constitution during this time.\n\n**Information 5**: This article is unrelated to the Indian Constitution and emergency provisions. It reports on a state of emergency declared in Colombia in 2020 in response to the COVID-19 pandemic, and the economic measures taken by the government to mitigate its impact.\n\nNow, to evaluate the <CLAIM>, we need to examine whether it is supported by the available information.\n\nThe <CLAIM> states that according to \"Article 116,\" a state of emergency will begin across Colombia at 6 a.m. on the 17th. However, none of the provided INFORMATION mentions Article 116 or supports this specific claim. Information 5 does report on a state of emergency declared in Colombia, but it does not mention Article 116 or a specific start time of 6 a.m. on the 17th.\n\nTherefore, based on the available information, I would classify the <CLAIM> as FALSE.", "response": "FALSE."}

print(x.keys())
print()
print('response:', x["response"])
print()
print(x["query"])


dict_keys(['query', 'response'])

response: FALSE.

Below is a CLAIM and the PRIOR KNOWLEDGE associated with it. Please classify the CLAIM as TRUE or FALSE based on the PRIOR KNOWLEDGE. If the content described by the CLAIM is correct, then classify it as TRUE; if the content described by the CLAIM is incorrect, then classify it as FALSE.

CLAIM: According to "Article 116," a state of emergency will begin across Colombia at 6 a.m. on the 17th.
Publication date: 2020-03-17

PRIOR KNOWLEDGE:
Information 1:
Publication date: None
Title: Emergency Provisions
Content:
If approved by both the houses, the Emergency continues for 6 months and can be extended to an indefinite period with an approval of the Parliament for every six months.
Suspension of Fundamental rights under Article 19: According to Article 358, when a proclamation of National Emergency is made, the six fundamental rights under article 19 are automatically suspended. Article 19 is automatically revived after the expiry of the

In [71]:
0.067 * 12192

816.864

In [74]:
(136 * 6 + 5) / 12192

0.06733923884514435

In [61]:
import numpy as np
from statsmodels.stats.inter_rater import fleiss_kappa

# 构建标注矩阵
data = np.array(
    [[2, 1], [1, 2], [2, 1], [2, 1], [2, 1], [2, 1]] * 136 + [[2, 1], [1, 2], [2, 1], [2, 1], [2, 1]] + [
    [3, 0],  # 三位标注者均选择第一类
    [3, 0],
    [3, 0],
    [3, 0],
    [0, 3],
    [3, 0],  # 三位标注者均选择第一类
]* 1895 + [[3, 0]]) 

# 计算 Fleiss' Kappa
kappa = fleiss_kappa(data, method='fleiss')
print("Fleiss' Kappa:", kappa)


Fleiss' Kappa: 0.8489833376129666


In [75]:
len(data)


12192

Existing misinformation detection benchmark datasets (e.g., COVMIS and LIAR2) are limited by their reliance on fact-checking labels that are prone to fac-tual inaccuracies due to cognitive con-straints of fact-checkers and outdated la-bels. Prior real-time misinformation de-tection tasks have been hindered by the dual problems of label redundancy and cold start. To this end, we propose a nov-el Cyclic Evidence-based Misinfor-mation Detection (CEMD) framework, which incorporates two core mechanisms: (i) a Retrieval Augmented Generation (RAG) pipeline that accesses the latest external knowledge to augment insuffi-cient prior knowledge, and (ii) a cyclic evidence-bootstrapping mechanism that mitigates label redundancy and cold start. We introduce an improved bench-mark dataset, COVMIS2, built upon COVMIS, and conduct comprehensive experiments to evaluate the efficacy of our framework. Our results demonstrate that CEMD outperforms the prior state-of-the-art (SOTA) baseline on LIAR2 by 11.95% and surpasses the human baseline on COVMIS2 by 6.31%, leveraging the Llama-3-70B-Instruct model to augment prior knowledge and the DoRA fine-tuned Llama-3-8B-Instruct model for bi-nary classification.

Existing misinformation detection benchmark datasets (e.g., COVMIS and LIAR2) are limited by their reliance on fact-checking labels that are prone to factual inaccuracies due to outdated la-bels and cognitive constraints of fact-checkers. Moreover, prior real-time mis-information detection tasks have been hindered by the dual problems of label redundancy and cold start. To this end, we propose a novel Cyclic Evidence-based Misinformation Detection (CEMD) framework, which incorporates two core mechanisms: (i) a Retrieval Augmented Generation (RAG) pipeline that accesses the latest external knowledge to augment insufficient prior knowledge, and (ii) a cyclic evidence-bootstrapping mecha-nism that mitigates label redundancy and cold start. We introduce an improved benchmark dataset, COVMIS2, built up-on COVMIS, and conduct comprehensive experiments to evaluate the efficacy of our framework. Our results demonstrate that CEMD outperforms the prior state-of-the-art (SOTA) baseline on LIAR2 by 11.95% and surpasses the human baseline on COVMIS2 by 6.31%, leveraging the Llama-3-70B-Instruct model to augment prior knowledge and the DoRA fine-tuned Llama-3-8B-Instruct model for bi-nary classification.


In [1]:
import torch
import torch.utils.benchmark as benchmark
from torch.backends.cuda import sdp_kernel, SDPBackend
# from torch.nn.attention import sdp_kernel
import torch.nn.functional as F
import os
import numpy as np
import random

DEFAULT_SEED = 42

def set_seed(seed=DEFAULT_SEED):
    torch.manual_seed(seed)
    torch.cuda.manual_seed_all(seed)
    np.random.seed(seed)
    random.seed(seed)
    torch.backends.cudnn.deterministic = True


# os.environ['CUDA_VISIBLE_DEVICES'] = '2'
# Lets define a helpful benchmarking function:
def benchmark_torch_function_in_microseconds(f, *args, **kwargs):
    t0 = benchmark.Timer(
        stmt="f(*args, **kwargs)", globals={"args": args, "kwargs": kwargs, "f": f}
    )
    return t0.blocked_autorange().mean * 1e6

# Lets define the hyper-parameters of our input
batch_size = 32
max_sequence_len = 1024
num_heads = 32
embed_dimension = 32

dtype = torch.float16
device = "cuda:0"

query = torch.rand(batch_size, num_heads, max_sequence_len, embed_dimension, device=device, dtype=dtype)
key = torch.rand(batch_size, num_heads, max_sequence_len, embed_dimension, device=device, dtype=dtype)
value = torch.rand(batch_size, num_heads, max_sequence_len, embed_dimension, device=device, dtype=dtype)

print(f"The default implementation runs in {benchmark_torch_function_in_microseconds(F.scaled_dot_product_attention, query, key, value):.3f} microseconds")

# Lets explore the speed of each of the 3 implementations

# Helpful arg mapper
backend_map = {
    SDPBackend.MATH: {"enable_math": True, "enable_flash": False, "enable_mem_efficient": False},
    SDPBackend.FLASH_ATTENTION: {"enable_math": False, "enable_flash": True, "enable_mem_efficient": False},
    SDPBackend.EFFICIENT_ATTENTION: {
        "enable_math": False, "enable_flash": False, "enable_mem_efficient": True}
}

with sdp_kernel(**backend_map[SDPBackend.MATH]):
    print(f"The math implementation runs in {benchmark_torch_function_in_microseconds(F.scaled_dot_product_attention, query, key, value):.3f} microseconds")

with sdp_kernel(**backend_map[SDPBackend.FLASH_ATTENTION]):
    try:
        print(f"The flash attention implementation runs in {benchmark_torch_function_in_microseconds(F.scaled_dot_product_attention, query, key, value):.3f} microseconds")
    except RuntimeError:
        print("FlashAttention is not supported. See warnings for reasons.")

with sdp_kernel(**backend_map[SDPBackend.EFFICIENT_ATTENTION]):
    try:
        print(f"The memory efficient implementation runs in {benchmark_torch_function_in_microseconds(F.scaled_dot_product_attention, query, key, value):.3f} microseconds")
    except RuntimeError:
        print("EfficientAttention is not supported. See warnings for reasons.")


The default implementation runs in 927.106 microseconds


/home/hanlv/miniconda3/envs/swift/lib/python3.11/site-packages/torch/backends/cuda/__init__.py:342: FutureWarning: torch.backends.cuda.sdp_kernel() is deprecated. In the future, this context manager will be removed. Please see, torch.nn.attention.sdpa_kernel() for the new context manager, with updated signature.
  warnings.warn(


The math implementation runs in 10094.952 microseconds
The flash attention implementation runs in 918.287 microseconds
The memory efficient implementation runs in 1782.254 microseconds


In [1]:
dtype = torch.bfloat16

set_seed()
query = torch.rand(batch_size, num_heads, max_sequence_len, embed_dimension, device=device, dtype=dtype)
key = torch.rand(batch_size, num_heads, max_sequence_len, embed_dimension, device=device, dtype=dtype)
value = torch.rand(batch_size, num_heads, max_sequence_len, embed_dimension, device=device, dtype=dtype)


# with sdp_kernel(**backend_map[SDPBackend.FLASH_ATTENTION]):
# print(F.scaled_dot_product_attention(query, key, value))
    


In [2]:
from decimal import Decimal
import decimal

def data_round(x):
    return str(Decimal(x).quantize(Decimal('0.0000'), rounding=decimal.ROUND_HALF_UP))


def res_trans(res):
    print(data_round(res["ACC"]))
    print(data_round(res["F1"]))
    print(data_round(res["Precision"]))
    print(data_round(res["Recall"]))

res = {
        "model": "Mistral-7B-Instruct-v0.3",
        "train_test_split": "8:2",
        "train_ratio": "1.0",
        "train_loss": 0.06997728,
        "lr": "1.4e-4",
        "ACC": 0.9877049180327869,
        "F1": 0.9832533980502562,
        "Precision": 0.9883868032359371,
        "Recall": 0.9783695652173913
    }

res_trans(res)


0.9877
0.9833
0.9884
0.9784


In [4]:
import numpy as np
x = np.array([0.9869,	0.9877,	0.9852,	0.9877,	0.9877])
x.mean()
# 0.98672 0.98622 0.9852399999999999

0.98704

In [10]:
import os
file = '/home/hanlv/workspace/code/research/infodemic/LLM/swift/examples/pytorch/llm/output/liar2/Llama-3-8B-Instruct/with_llama3_info/brave/data1a-split=8:1:1-ratio=1.0/dora-r=8/lr=1.1e-4-20240827-13:17:06'



get_ckpts(file)

['checkpoint-1148']